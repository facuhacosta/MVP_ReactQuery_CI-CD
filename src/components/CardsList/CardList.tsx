import { useQuery } from "@tanstack/react-query"
import { fetchUsers } from "../../assets/utils"
import Card from "../Card/Card"
import styles from "./CardList.module.scss"

const CardList = () => {
  const { isLoading, isError, data: characters = [] } = useQuery({
    queryKey: ['characters'],
    queryFn: async () => await fetchUsers()
  })

  return (
    <div className={styles.container}>
      {characters.length > 0 && characters.map((character: Character) => (
        <Card character={character} />
      ))}

      {isLoading && <p>Is Loading ...</p>}
      {isError && <p>Error en la Carga</p>}
      {!isError && !isLoading && characters.length === 0 && <p>No hay Usuarios</p>}
    </div>
  )
}

export default CardList