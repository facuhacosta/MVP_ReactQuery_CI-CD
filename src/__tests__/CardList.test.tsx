import CardList from "../components/CardsList/CardList"
import { render } from "@testing-library/react"
import { QueryClientProvider, QueryClient } from '@tanstack/react-query'
import { expect, describe, it, vi, afterEach } from "vitest"


describe("App test", () => {

  const mocks = vi.hoisted(() => ({
    characters: [
      {
        "name": "3-D Man",
        "thumbnail": {
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784",
          "extension": "jpg"
        },
        "comics": {
          "items": [
            {
              "name": "Avengers: The Initiative (2007) #14"
            },
            {
              "name": "Avengers: The Initiative (2007) #15"
            },
            {
              "name": "Marvel Premiere (1972) #35"
            }
          ]
        }
      }
    ]
  }))

  afterEach(() => {
    vi.clearAllMocks()
  })

  it("Sould Render Character", async () => {

    vi.mock('@tanstack/react-query', async () => {
      const actual: {} = await vi.importActual('@tanstack/react-query')
      return {
        ...actual,
        useQuery: vi.fn().mockReturnValueOnce({ data: mocks.characters, isLoading: false, error: {} }),
      }
    })

    const queryClient = new QueryClient()

    const { asFragment, findByText } = render(
      <QueryClientProvider client={queryClient}>
        <CardList />
      </QueryClientProvider>
    )
    await findByText('3-D Man')
    expect(asFragment()).toMatchSnapshot()
  })

  it("Sould show loading", async () => {

    vi.mock('@tanstack/react-query', async () => {
      const actual: {} = await vi.importActual('@tanstack/react-query')
      return {
        ...actual,
        useQuery: vi.fn().mockReturnValue({ data: mocks.characters, isLoading: true, error: {} }),
      }
    })

    const queryClient = new QueryClient()

    const { asFragment, findAllByText } = render(
      <QueryClientProvider client={queryClient}>
        <CardList />
      </QueryClientProvider>
    )
    await findAllByText('Is Loading ...')
    expect(asFragment()).toMatchSnapshot()
  })
})