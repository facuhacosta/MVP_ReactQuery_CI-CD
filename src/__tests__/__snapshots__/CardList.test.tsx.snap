// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App test > Sould Render Character 1`] = `
<DocumentFragment>
  <div
    class="_container_5ea73b"
  >
    <div
      class="_container_d75732"
    >
      <div
        class="_imageContainer_d75732"
      >
        <img
          alt=""
          src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
        />
      </div>
      <section>
        <h3>
          3-D Man
        </h3>
      </section>
      <ol>
        <li>
          Avengers: The Initiative (2007) #14
        </li>
        <li>
          Avengers: The Initiative (2007) #15
        </li>
        <li>
          Marvel Premiere (1972) #35
        </li>
      </ol>
    </div>
    <p>
      Is Loading ...
    </p>
  </div>
</DocumentFragment>
`;

exports[`App test > Sould show loading 1`] = `
<DocumentFragment>
  <div
    class="_container_5ea73b"
  >
    <div
      class="_container_d75732"
    >
      <div
        class="_imageContainer_d75732"
      >
        <img
          alt=""
          src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
        />
      </div>
      <section>
        <h3>
          3-D Man
        </h3>
      </section>
      <ol>
        <li>
          Avengers: The Initiative (2007) #14
        </li>
        <li>
          Avengers: The Initiative (2007) #15
        </li>
        <li>
          Marvel Premiere (1972) #35
        </li>
      </ol>
    </div>
    <p>
      Is Loading ...
    </p>
  </div>
</DocumentFragment>
`;
